stake-pool/single-pool-cli/src/cli.rs

```diff
@@ -0,0 +1,399 @@
+    #[clap(long, id = "AUTHORIZED_WITHDRAWER_KEYPAIR", validator = |s| is_valid_signer(s))]
+    pub authorized_withdrawer: Option<SignerArg>,
+}
+
+#[derive(Clone, Debug, Args)]
+#[clap(group(ArgGroup::new("pool-source").required(true).args(&["pool-address", "vote-account-address"])))]
+pub struct CreateStakeCli {
+    /// Number of lamports to stake
+    pub lamports: u64,
+
+    /// The pool to create a stake account for
+    #[clap(short, long = "pool", value_parser = Pubkey::from_str)]
+    pub pool_address: Option<Pubkey>,
+
+    /// The vote account corresponding to the pool to create stake for
+    #[clap(long = "vote-account", value_parser = Pubkey::from_str)]
+    pub vote_account_address: Option<Pubkey>,
+
+    /// Authority to assign to the new stake account. Defaults to the pubkey of the client keypair
+    #[clap(long = "stake-authority", value_parser = Pubkey::from_str)]
+    pub stake_authority_address: Option<Pubkey>,
+}
+
+#[derive(Clone, Debug, Args)]
+#[clap(group(ArgGroup::new("pool-source").required(true).args(&["pool-address", "vote-account-address"])))]
+pub struct DisplayCli {
+    /// The pool to display
+    #[clap(value_parser = Pubkey::from_str)]
+    pub pool_address: Option<Pubkey>,
+
+    /// The vote account corresponding to the pool to display
+    #[clap(long = "vote-account", value_parser = Pubkey::from_str)]
+    pub vote_account_address: Option<Pubkey>,
+}
+
+pub fn parse_json_rpc_url(url_or_moniker: &str) -> Result<String, String> {
+    is_url_or_moniker(url_or_moniker).map(|_| normalize_to_url_if_moniker(url_or_moniker))
+}
+
+pub fn parse_output_format(output_format: &str) -> OutputFormatArg {
+    match output_format {
+        "json" => OutputFormatArg::Json,
+        "json-compact" => OutputFormatArg::JsonCompact,
+        _ => unreachable!(),
+    }
+}
+
+pub fn is_valid_token_name(s: &str) -> Result<(), String> {
+    if s.len() > 32 {
+        Err("Maximum token name length is 32 characters".to_string())
+    } else {
+        Ok(())
+    }
+}
+
+pub fn is_valid_token_uri(s: &str) -> Result<(), String> {
+    if s.len() > 10 {
+        Err("Maximum token symbol length is 10 characters".to_string())
+    } else {
+        Ok(())
+    }
+}
+
+pub fn pool_address_from_args(maybe_pool: Option<Pubkey>, maybe_vote: Option<Pubkey>) -> Pubkey {
+    if let Some(pool_address) = maybe_pool {
+        pool_address
+    } else if let Some(vote_account_address) = maybe_vote {
+        find_pool_address(&single_pool::id(), &vote_account_address)
+    } else {
+        unreachable!()
+    }
+}
+
+// this plus with_signer(s) is because solana clap v3 utils requires ArgMatches
+// which means its impossible to parse keypairs in value_parser
+// instead, we take the input into a string wrapper from the cli
+// and then once the first pass is over, we do a second manual pass converting to signer wrappers
+#[derive(Clone, Debug)]
+pub enum SignerArg {
+    Source(String),
+    Signer(Arc<dyn Signer>),
+}
```

Comment:
I believe this is a longstanding issue with why the conversion to clap v3 will be so difficult. Maybe @samkim-crypto can weigh in since he's been doing work in that area

Reply:
looking at this again i think the way forward is writing a reduced functionality version that eliminates the four uses of `ArgMatches`. i was gonna kick the can but while writing the issue i decided its actually easy after all lol

it will be in 1.17 when i pr it so the `SignerArg` stuff needs to stay for now but this is definitely temporary

https://github.com/solana-labs/solana/issues/32421

Reply:
this turns out to be nontrivial and maybe not possible, because `Signer` cant require `Send + Sync` because of the c wrapper library necessary for ledger support

Reply:
We might be able to find a way around that by wrapping in `Arc`s or `AtomicPtr`s or something... which type is causing the issue? Is it `RemoteWalletType`?

Reply:
`RemoteKeypair`, have more detail in https://github.com/solana-labs/solana/issues/32421